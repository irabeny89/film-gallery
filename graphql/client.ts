import { ApolloClient, InMemoryCache } from "@apollo/client";
import { RestLink } from "apollo-link-rest";
import config from "config";

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: new RestLink({
    uri: `https://www.omdbapi.com/?apikey=${config.apiKey}&`,
    endpoints: { poster: `https://img.omdbapi.com/?apikey=${config.apiKey}&` },
  }),
});

export default client;
