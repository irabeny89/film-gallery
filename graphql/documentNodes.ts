import { gql } from "@apollo/client";

export const SEARCH_TITLE = gql`
  query SearchTitle($title: String!) {
    film(title: $title) @rest(type: "Film", path: "s={args.title}") {
      Search
    }
  }
`