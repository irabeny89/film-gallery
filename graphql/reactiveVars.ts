import { makeVar } from "@apollo/client";
import mockSearchResult from "data/mockSearchResult";
import { FilmSearchResultType } from "types";
import config from "config";

const initialSearchResult = {
    Search: [],
  },
  defaultSearchResult =
    config.network === "offline" ? mockSearchResult : initialSearchResult;

export const filmSearchResultVar =
  makeVar<FilmSearchResultType>(defaultSearchResult);
