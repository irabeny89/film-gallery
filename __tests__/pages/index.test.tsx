import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import HomePage from "pages/index";
import { FilmSearchResultType } from "types";
import { SEARCH_TITLE } from "graphql/documentNodes";
import mockSearchResult from "data/mockSearchResult";

const searchTerm = "avengers",
  mocks: readonly MockedResponse<Record<"film", FilmSearchResultType>>[] = [
    {
      request: {
        query: SEARCH_TITLE,
        variables: {
          title: searchTerm,
        },
      },
      result: {
        data: {
          film: mockSearchResult,
        },
      },
    },
  ];

describe("Home Page", () => {
  beforeEach(() => {
    render(
      <MockedProvider mocks={mocks}>
        <HomePage />
      </MockedProvider>
    );
  });
  afterEach(cleanup);

  it("has a page title.", () => {
    const pageTitle = screen.getByRole("heading");

    expect(pageTitle).toHaveTextContent("Welcome to the film gallery");
  });

  it("list nine films.", async () => {
    const filmList = await screen.findByTestId("film-list"),
      filmCards = await screen.findAllByTestId("film-card");

    expect(filmList).not.toBeEmptyDOMElement();
    expect(filmCards.length).toBe(9);
  });

  it("does not search when textbox is empty.", async () => {
    const searchButton = screen.getByRole("button"),
      textbox = screen.getByRole("textbox"),
      resultModal = screen.queryByTestId("result-modal"),
      user = userEvent.setup();

    expect(textbox).toHaveValue("");
    await user.click(searchButton);
    expect(resultModal).not.toBeInTheDocument();
  });

  it("searches and returns list of avengers films.", async () => {
    const searchButton = screen.getByRole("button"),
      textbox = screen.getByRole("textbox"),
      user = userEvent.setup();

    await user.type(textbox, searchTerm), await user.click(searchButton);

    expect(Array.isArray(await screen.findAllByText(/avengers/i))).toBe(true);
  });

  it("matches snapshot.", () => {
    const mainElement = screen.getByRole("main");

    expect(mainElement).toMatchSnapshot();
  });
});
