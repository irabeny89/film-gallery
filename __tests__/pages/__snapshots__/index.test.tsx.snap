// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page matches snapshot. 1`] = `
<main
  class="main"
>
  <svg
    color="red"
    fill="currentColor"
    height="100"
    stroke="currentColor"
    stroke-width="0"
    style="color: red;"
    viewBox="0 0 20 20"
    width="100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clip-rule="evenodd"
      d="M4 3a2 2 0 00-2 2v10a2 2 0 002 2h12a2 2 0 002-2V5a2 2 0 00-2-2H4zm3 2h6v4H7V5zm8 8v2h1v-2h-1zm-2-2H7v4h6v-4zm2 0h1V9h-1v2zm1-4V5h-1v2h1zM5 5v2H4V5h1zm0 4H4v2h1V9zm-1 4h1v2H4v-2z"
      fill-rule="evenodd"
    />
  </svg>
  <h1
    class="title"
  >
    Welcome to the film gallery
  </h1>
  <p
    class="description"
  >
    Search for your 
    <i>
      favorite
    </i>
     
    <code
      class="code"
    >
      <svg
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4 3a2 2 0 00-2 2v10a2 2 0 002 2h12a2 2 0 002-2V5a2 2 0 00-2-2H4zm3 2h6v4H7V5zm8 8v2h1v-2h-1zm-2-2H7v4h6v-4zm2 0h1V9h-1v2zm1-4V5h-1v2h1zM5 5v2H4V5h1zm0 4H4v2h1V9zm-1 4h1v2H4v-2z"
          fill-rule="evenodd"
        />
      </svg>
      film
    </code>
  </p>
  <div
    class="mb-5 container-fluid"
  >
    <div
      class="toast-container position-absolute top-0 start-50 translate-middle-x"
    />
    <div
      class="justify-content-center row"
    >
      <div
        class="col-sm-10"
      >
        <form
          class=""
        >
          <div
            class="justify-content-center input-group"
          >
            <div
              class="form-floating"
            >
              <input
                arial-label="search GitHub repositories"
                class="form-control"
                name="search"
                placeholder="Enter movie title"
              />
              <label>
                Enter movie title
              </label>
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              <svg
                fill="none"
                height="25"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="25"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M18.319 14.4326C20.7628 11.2941 20.542 6.75347 17.6569 3.86829C14.5327 0.744098 9.46734 0.744098 6.34315 3.86829C3.21895 6.99249 3.21895 12.0578 6.34315 15.182C9.22833 18.0672 13.769 18.2879 16.9075 15.8442C16.921 15.8595 16.9351 15.8745 16.9497 15.8891L21.1924 20.1317C21.5829 20.5223 22.2161 20.5223 22.6066 20.1317C22.9971 19.7412 22.9971 19.1081 22.6066 18.7175L18.364 14.4749C18.3493 14.4603 18.3343 14.4462 18.319 14.4326ZM16.2426 5.28251C18.5858 7.62565 18.5858 11.4246 16.2426 13.7678C13.8995 16.1109 10.1005 16.1109 7.75736 13.7678C5.41421 11.4246 5.41421 7.62565 7.75736 5.28251C10.1005 2.93936 13.8995 2.93936 16.2426 5.28251Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    class="spinner-border"
  />
</main>
`;
